File: 프로그래머스/lv4/43236. 징검다리/징검다리.js
function solution(distance, rocks, n) {
  rocks = rocks.sort((a, b) => a - b);
  let subDistances = Array.from({ length: rocks.length + 1 }, () => 0);
  subDistances[0] = rocks[0];
  subDistances[rocks.length] = distance - rocks[rocks.length - 1];
  for (let i = 1; i < rocks.length; i++)
    subDistances[i] = rocks[i] - rocks[i - 1];
  let left = 1;
  let right = distance;
  let mid = Math.floor((left + right) / 2);
  let current, removed;
  while (right - left > 1) {
    removed = 0;
    current = 0;
    for (let i = 0; i < subDistances.length; i++) {
      current += subDistances[i];
      if (current < mid) removed++;
      else current = 0;
    }
    if (removed > n) right = mid;
    else left = mid;
    mid = Math.floor((left + right) / 2);
  }
  return mid;
}

File: 프로그래머스/4/43236. 징검다리/징검다리.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool isPossible(long long mid, const vector<int> &rocks, int n, int distance)
{
    int removed = 0;
    long long previous = 0;

    for (auto rock : rocks)
    {
        if ((rock - previous) < mid)
        {

            removed++;
            if (removed > n)
                return false;
        }
        else
        {

            previous = rock;
        }
    }

    if ((distance - previous) < mid)
    {
        removed++;
    }

    return removed <= n;
}

long long solution(int distance, vector<int> rocks, int n)
{

    sort(rocks.begin(), rocks.end());

    long long left = 1;
    long long right = distance;
    long long answer = 0;

    while (left <= right)
    {
        long long mid = (left + right) / 2;

        if (isPossible(mid, rocks, n, distance))
        {

            answer = mid;
            left = mid + 1;
        }
        else
        {

            right = mid - 1;
        }
    }

    return answer;
}